feat: warn by default for unused disable directives

bmish opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Fixes #15466. This is part two of that issue. Part one was in #17212.

The goal here is to warn by default for reporting unused disable directives, only when using the new flat config.

TODO:
- [x] Fix 35 tests in `tests/lib/linter/linter.js`
- [x] Update docs
- [x] Add new test

#### Is there anything you'd like reviewers to focus on?

1. Is this the right place to change the default?
2. Where does the default need to be documented?

<!-- markdownlint-disable-file MD004 -->

mdjermanovic commented:
> 1. Is this the right place to change the default?

I was thinking about setting this default in [`default-config.js`](https://github.com/eslint/eslint/blob/main/lib/config/default-config.js). For example, in its first object. That way, it will appear in `--print-config`, which is good I believe.

```diff
     languageOptions: {
         sourceType: "module",
         ecmaVersion: "latest",
         parser: require("espree"),
         parserOptions: {}
     },
+    linterOptions: {
+        reportUnusedDisableDirectives: 1
+    }
```

mdjermanovic commented:
@bmish are you still working on this? This is one of the last changes planned for the upcoming v9.0.0-alpha.0.

> * Fix 35 tests in `tests/lib/linter/linter.js`

Could we maybe fix those tests by adding `linterOptions: { reportUnusedDisableDirectives: 0 }` to the configs they're passing to `linter.verify()`?

mdjermanovic commented:
> 2\. Where does the default need to be documented?

I think here:

https://eslint.org/docs/head/use/configure/configuration-files#reporting-unused-disable-directives

Also here (at some point in the future, we'll maybe clear this redundancy we now have in the docs):

https://eslint.org/docs/head/use/configure/rules#report-unused-eslint-disable-comments



And maybe here, at the end of the Default Value line we could add `(defaults to "warn")`:

https://eslint.org/docs/head/use/command-line-interface#--report-unused-disable-directives-severity

bmish commented:
I moved the default to `default-config.js` and fixed the existing tests. Thanks for the tips.

I still need to add a test for the new default and update some more docs. Will try to do this today.

bmish commented:
I have updated the docs and added tests. This is ready for review.

nzakas commented on tests/lib/rules/capitalized-comments.js:
```diff
@@ -59,8 +59,8 @@ ruleTester.run("capitalized-comments", rule, {
         // No options: eslint/istanbul/jshint/jscs/globals?/exported are okay
```
I think in this and the other rule tests, the intent was to check if directive comments were handled correctly.

Rather than changing all directive comments to non-directive comments, could we instead just turn off reporting of unused directives in those files?

mdjermanovic replied:
Looks like the old rule tests aren't actually failing now, because FlatRuleTester has its own `baseConfig`, where `linterOptions.reportUnusedDisableDirectives` isn't set.

https://github.com/eslint/eslint/blob/d1018fc5e59db0495aa4a7f501c9d3f831981f35/lib/rule-tester/flat-rule-tester.js#L538-L586

Now the question is shall we leave it that way?

bmish replied:
Good points. I reverted the rule tests.

On the one hand, it seems like rule tests should generally use the default config? But on the other hand, reporting disable directives seems like it would be unnecessary and even annoying during rule tests, so I'd probably prefer to leave that option off...

nzakas replied:
I think reporting unused directives is probably best left off for rule tests -- I can't imagine people wanting those warnings intermixed with rule errors.

nzakas reviewed (changes requested):
Overall LGTM. Just a note on changing rule tests.

nzakas reviewed (approved):
LGTM. Would like @mdjermanovic to confirm before merging.

mdjermanovic reviewed (approved):
LGTM, thanks!
